File: examples/csharp/Scripts/PullToRefreshScene.cs
using Godot;
using System.Threading.Tasks;

namespace FairyGUI.Examples;

public partial class PullToRefreshScene : DemoSceneBase
{
    private GuiObject _list1;
    private GuiObject _list2;
    private bool _refreshing1;
    private bool _refreshing2;

    protected override void ContinueInit()
    {
        UIPackage.AddPackage("res://Resources/UI/PullToRefresh");
        _view = UIPackage.CreateObject("PullToRefresh", "Main");
        _groot.AddChild(_view);
        CallDeferred(MethodName.SetupLists);
    }

    private void SetupLists()
    {
        if (!IsUiActive())
            return;

        _list1 = _view.GetChild("list1");
        if (_list1 != null)
        {
            _list1.SetItemRenderer(new Callable(this, MethodName.RenderListItem1));
            _list1.SetVirtual();
            _list1.SetNumItems(1);
            _list1.AddEventListener(FguiEvent.PullDownRelease, new Callable(this, MethodName.OnPullDownRelease));
        }

        _list2 = _view.GetChild("list2");
        if (_list2 != null)
        {
            _list2.SetItemRenderer(new Callable(this, MethodName.RenderListItem2));
            _list2.SetVirtual();
            _list2.SetNumItems(1);
            _list2.AddEventListener(FguiEvent.PullUpRelease, new Callable(this, MethodName.OnPullUpRelease));
        }
    }

    private void RenderListItem1(int index, GodotObject obj)
    {
        var num = _list1.GetNumItems();
        obj.SetText("Item " + (num - index - 1));
    }

    private void RenderListItem2(int index, GodotObject obj)
    {
        obj.SetText("Item " + index);
    }

    private void OnPullDownRelease(FguiEventContext ctx)
    {
        if (!IsUiActive() || _refreshing1)
            return;
        CallDeferred(MethodName.BeginPullDownRefresh);
    }

    private void OnPullUpRelease(FguiEventContext ctx)
    {
        if (!IsUiActive() || _refreshing2)
            return;
        CallDeferred(MethodName.BeginPullUpRefresh);
    }

    private async void BeginPullDownRefresh()
    {
        if (_refreshing1 || !IsUiActive() || _list1 == null)
            return;
        _refreshing1 = true;
        var sp = _list1.GetScrollPane();
        if (sp == null)
        {
            _refreshing1 = false;
            return;
        }

        var header = sp.GetHeader();
        if (header != null)
        {
            var c1 = header.GetController("c1");
            c1?.SetSelectedIndex(2);
            sp.LockHeader((int)header.GetHeight());
        }

        await WaitSeconds(2.0);
        if (!IsUiActive() || _list1 == null)
        {
            _refreshing1 = false;
            return;
        }

        _list1.SetNumItems(_list1.GetNumItems() + 5);

        if (header != null)
        {
            var c1 = header.GetController("c1");
            c1?.SetSelectedIndex(3);
            sp.LockHeader(35);
        }

        await WaitSeconds(2.0);
        if (!IsUiActive() || _list1 == null)
        {
            _refreshing1 = false;
            return;
        }

        if (header != null)
        {
            var c1 = header.GetController("c1");
            c1?.SetSelectedIndex(0);
            sp.LockHeader(0);
        }
        _refreshing1 = false;
    }

    private async void BeginPullUpRefresh()
    {
        if (_refreshing2 || !IsUiActive() || _list2 == null)
            return;
        _refreshing2 = true;
        var sp = _list2.GetScrollPane();
        if (sp == null)
        {
            _refreshing2 = false;
            return;
        }

        var footer = sp.GetFooter();
        if (footer != null)
        {
            var c1 = footer.GetController("c1");
            c1?.SetSelectedIndex(1);
            sp.LockFooter((int)footer.GetHeight());
        }

        await WaitSeconds(2.0);
        if (!IsUiActive() || _list2 == null)
        {
            _refreshing2 = false;
            return;
        }

        _list2.SetNumItems(_list2.GetNumItems() + 5);

        if (footer != null)
        {
            var c1 = footer.GetController("c1");
            c1?.SetSelectedIndex(0);
            sp.LockFooter(0);
        }
        _refreshing2 = false;
    }
}

File: examples/ts/Scripts/PullToRefreshScene.ts
/// <reference path="../fairygui.d.ts" />

import { Callable } from "godot";
import { DemoSceneBase } from "./DemoSceneBase";

export default class PullToRefreshScene extends DemoSceneBase {
    private _list1: GList | null = null;
    private _list2: GList | null = null;
    private _refreshing1 = false;
    private _refreshing2 = false;

    protected ContinueInit(): void {
        UIPackage.addPackage("res://Resources/UI/PullToRefresh");
        this._view = UIPackage.createObject("PullToRefresh", "Main");
        this._groot!.addChild(this._view);

        this._list1 = this._view.getChild("list1") as GList | null;
        if (this._list1 != null) {
            this._list1.setItemRenderer(this._renderListItem1.bind(this));
            this._list1.setVirtual();
            this._list1.setNumItems(1);
            this._list1.addEventListener(
                UIEventDispatcher.PULLDOWNRELEASE,
                Callable.create(this._onPullDownToRefresh.bind(this)),
            );
        }

        this._list2 = this._view.getChild("list2") as GList | null;
        if (this._list2 != null) {
            this._list2.setItemRenderer(this._renderListItem2.bind(this));
            this._list2.setVirtual();
            this._list2.setNumItems(1);
            this._list2.addEventListener(
                UIEventDispatcher.PULLUPRELEASE,
                Callable.create(this._onPullUpToRefresh.bind(this)),
            );
        }
    }

    private _renderListItem1(index: number, obj: GComponent): void {
        const num = this._list1!.getNumItems();
        obj.setText(`Item ${String(num - index - 1)}`);
    }

    private _renderListItem2(index: number, obj: GComponent): void {
        obj.setText(`Item ${String(index)}`);
    }

    private async _onPullDownToRefresh(): Promise<void> {
        if (this._refreshing1 || this._list1 == null) {
            return;
        }
        this._refreshing1 = true;
        const sp = this._list1.getScrollPane();
        if (sp == null) {
            this._refreshing1 = false;
            return;
        }
        const header = sp.getHeader();
        if (header != null) {
            const c1 = header.getController("c1");
            c1!.selectedIndex = 2;
            sp.lockHeader(header.getHeight());
        }

        await this.waitSeconds(2.0);
        if (!this.isSceneActive() || this._list1 == null) {
            this._refreshing1 = false;
            return;
        }

        this._list1.setNumItems(this._list1.getNumItems() + 5);

        if (header != null) {
            const c1 = header.getController("c1");
            c1!.selectedIndex = 3;
            sp.lockHeader(35);
        }

        await this.waitSeconds(2.0);
        if (!this.isSceneActive() || this._list1 == null) {
            this._refreshing1 = false;
            return;
        }

        if (header != null) {
            const c1 = header.getController("c1");
            c1!.selectedIndex = 0;
            sp.lockHeader(0);
        }
        this._refreshing1 = false;
    }

    private async _onPullUpToRefresh(): Promise<void> {
        if (this._refreshing2 || this._list2 == null) {
            return;
        }
        this._refreshing2 = true;
        const sp = this._list2.getScrollPane();
        if (sp == null) {
            this._refreshing2 = false;
            return;
        }
        const footer = sp.getFooter();
        if (footer != null) {
            const c1 = footer.getController("c1");
            c1!.selectedIndex = 1;
            sp.lockFooter(footer.getHeight());
        }

        await this.waitSeconds(2.0);
        if (!this.isSceneActive() || this._list2 == null) {
            this._refreshing2 = false;
            return;
        }

        this._list2.setNumItems(this._list2.getNumItems() + 5);

        if (footer != null) {
            const c1 = footer.getController("c1");
            c1!.selectedIndex = 0;
            sp.lockFooter(0);
        }
        this._refreshing2 = false;
    }
}
